src/lib.rs

```diff
@@ -56,12 +57,28 @@ macro_rules! impl_array {
                 mut topic_parts: core::iter::Peekable<core::str::Split<char>>,
                 value: &[u8],
             ) -> Result<(), Error> {
-                if topic_parts.peek().is_some() {
-                    return Err(Error::NameTooLong);
+                if let Some(next) = topic_parts.next() {
+                    // Parse what should be the index value
+                    let i: usize = serde_json_core::from_str(next)?.0;
+
+                    // There should not be any more topic parts after the index
+                    if topic_parts.peek().is_some() {
```

Comment:
This isn't necessarily true. What if it's an array of structures, each of which implement `StringSet`?